Delay shard reassignment from nodes which are known to be restarting

AthenaEryma opened the pull request:
This PR makes the delayed allocation infrastructure aware of registered node shutdowns, so that reallocation of shards will be further delayed for nodes which are known to be restarting.

To make this more configurable, the Node Shutdown APIs now support a `allocation_delay` parameter, which defaults to 5 minutes. For example:
```
PUT /_nodes/USpTGYaBSIKbgSUJR2Z9lg/shutdown
{
  "type": "restart",
  "reason": "Demonstrating how the node shutdown API works",
  "allocation_delay": "20m"
}
```

Will cause reallocation of shards assigned to that node to another node to be delayed by 20 minutes. Note that this delay will only be used if it's *longer* than the index-level allocation delay, set via `index.unassigned.node_left.delayed_timeout`.

The `allocation_delay` parameter is only valid for `restart`-type shutdown registrations, and the request will be rejected if it's used with another shutdown type.

Relates https://github.com/elastic/elasticsearch/issues/70338

AthenaEryma committed:
First cut of an integration test (fails, obviously)

AthenaEryma committed:
WIP - test passes!

AthenaEryma committed:
Clean up getRemainingDelay params + reinstate unit test

AthenaEryma committed:
Back to taking a `Metadata` since we need a setting...

AthenaEryma committed:
Add a bunch of unit tests

AthenaEryma committed:
Add shard allocation delay to `SingleNodeShutdownMetadata`

AthenaEryma committed:
Value plumbed into getRemainingDelay + more tests

AthenaEryma committed:
spotless

AthenaEryma committed:
Imports

AthenaEryma committed:
EMPTY_MAP -> emptyMap()

AthenaEryma committed:
More integration tests

AthenaEryma committed:
spotless

AthenaEryma committed:
Fix compilation in ML tests

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

gwbrown commented:
@henningandersen I've pinged you for a review on this PR mostly because I don't know who would be most appropriate from the Distributed team to review this - feel free to reassign as you see fit.

gwbrown commented:
Failure is https://github.com/elastic/elasticsearch/issues/75667.

@elasticmachine run elasticsearch-ci/part-1

gwbrown commented:
@DaveCTurner I've pinged you for a review on this PR mostly because I don't know who would be most appropriate from the Distributed team to review this - feel free to reassign as you see fit.

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/SingleNodeShutdownMetadata.java:
```diff
@@ -217,15 +252,25 @@ public Builder setStartedAtMillis(long startedAtMillis) {
                 throw new IllegalArgumentException("start timestamp must be set");
             }
+            TimeValue delayOrDefault = shardReallocationDelay;
+            if (Type.RESTART.equals(type) && delayOrDefault == null) {
+                delayOrDefault = DEFAULT_RESTART_SHARD_ALLOCATION_DELAY;
```
This feels like the wrong place to implement this default, I think rather than here, it should go into the getter inside of `SingleNodeShutdownMetadata`, otherwise calling `toXContent` on the metadata makes it appear that it has been set (when in reality the default value will be used)

gwbrown replied:
Good call, I'll change it.

dakrone commented on server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java:
```diff
@@ -9,36 +9,42 @@
  */
 public final class UnassignedInfo implements ToXContentFragment, Writeable {
 
+    private static final Version LAST_ALLOCATED_NODE_VERSION = Version.V_8_0_0;
```
This feels a little strange to me, do we need this rather than just using the version directly in the two serialization methods?

gwbrown replied:
I like doing this because it 1) adds semantic information to the version check, and 2) is less error-prone when switching the serialization version following a backport, especially if there are other serialization version checks in the same class.

I can make it inline if you feel strongly about it.

dakrone replied:
I think it's fine to keep it, but I would suggest two things. First, we should add a comment about what it is for, and second, I think renaming it to something like `VERSION_LAST_ALLOCATED_NODE_ADDED` would be better.

Just looking at it with its current name makes me think this is the version of the *node* where the shard was last allocated.

dakrone commented on server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java:
```diff
@@ -366,8 +390,18 @@ public AllocationStatus getLastAllocationStatus() {
+        final Settings indexSettings,
+        final Map<String, SingleNodeShutdownMetadata> nodesShutdownMap
+    ) {
+        Map<String, SingleNodeShutdownMetadata> nodeShutdowns = nodesShutdownMap != null ? nodesShutdownMap : Collections.emptyMap();
```
I don't think we need this `null` check here, since `Metadata#nodeShutdowns()` returns an empty map if there are no shutdowns. Maybe we can replace it with an `assert` instead?

gwbrown replied:
👍 I think this is left over from an intermediate version.

dakrone commented on x-pack/plugin/shutdown/src/internalClusterTest/java/org/elasticsearch/xpack/shutdown/NodeShutdownDelayedAllocationIT.java:
```diff
@@ -0,0 +1,241 @@
+            assertTrue(putShutdownResponse.isAcknowledged());
+        }
+
+        // Verify that the shard's allocation is still delayed
+        assertBusy(
+            () -> { assertThat(client().admin().cluster().prepareHealth().get().getDelayedUnassignedShards(), equalTo(1)); },
+            2,
+            TimeUnit.SECONDS
+        );
```
I think we should remove this assert, since a long GC or other CI slowness could cause this to be flaky. I think just the `ensureGreen` below is enough, because it was changed from 3 hours to 10 seconds and the change took effect.

dakrone commented on x-pack/plugin/shutdown/src/internalClusterTest/java/org/elasticsearch/xpack/shutdown/NodeShutdownDelayedAllocationIT.java:
```diff
@@ -0,0 +1,241 @@
+        // Verify that the shard's allocation is delayed - but with a shorter wait than the reallocation timeout
+        assertBusy(
+            () -> { assertThat(client().admin().cluster().prepareHealth().get().getDelayedUnassignedShards(), equalTo(1)); },
+            2,
+            TimeUnit.SECONDS
```
I think we can remove the 2 second limit here, since we have 3 hours to check this :)

dakrone reviewed (changes requested):
Thanks for working on this Gordon! I left some pretty minor comments, but nothing major

AthenaEryma committed:
Change version constant name & add comment per review

AthenaEryma committed:
Remove unnecessary null check per review

AthenaEryma committed:
Remove potentially-flaky assert per review.

AthenaEryma committed:
Remove timeout override on assert per review

AthenaEryma committed:
Move default reallocation delay to getting per review

AthenaEryma committed:
Remove unnecessary unit test

AthenaEryma committed:
Rename + change incorrect comment on unit test

dakrone reviewed (approved):
LGTM!

gwbrown commented:
@elasticmachine update branch

gwbrown commented:
@DaveCTurner Would you prefer us to wait for a review from the Distributed team on this, or should we go ahead and merge?

henningandersen commented on x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/PutShutdownNodeAction.java:
```diff
@@ -85,6 +94,10 @@ public String getReason() {
+        }
+
         @Override
         public ActionRequestValidationException validate() {
```
Perhaps add the validation of type vs reallocation delay here?

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java:
```diff
@@ -366,8 +394,17 @@ public AllocationStatus getLastAllocationStatus() {
+    ) {
+        long delayTimeoutNanos = Optional.ofNullable(lastAllocatedNodeId)
+            .map(nodesShutdownMap::get)
+            .filter(shutdownMetadata -> SingleNodeShutdownMetadata.Type.RESTART.equals(shutdownMetadata.getType()))
```
It is a bit odd that if a node crashes and then has a shutdown indication added, we extend the node left timeout to be the possibly longer shutdown node left timeout.

I wonder if we could instead capture whether the node was ready for shutdown when we lost it and use that to determine which of the delays to use?

gwbrown replied:
See my above comments for the difficulty with doing so, but as an alternative, we could at least capture whether the node was registered for shutdown at all when the node went offline.

henningandersen replied:
Yes, given your earlier comments, that approach makes sense to me.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/SingleNodeShutdownMetadata.java:
```diff
@@ -35,14 +38,16 @@
     public static final ParseField REASON_FIELD = new ParseField("reason");
     public static final String STARTED_AT_READABLE_FIELD = "shutdown_started";
     public static final ParseField STARTED_AT_MILLIS_FIELD = new ParseField(STARTED_AT_READABLE_FIELD + "millis");
+    public static final ParseField SHARD_REALLOCATION_DELAY = new ParseField("shard_reallocation_delay");
```
Perhaps this argument could be just "allocation_delay"?

henningandersen reviewed (commented):
I left a number of comments. I would like your input on whether we could instead capture which delay to use when we lose the node (see comments inline)?

I did not find any shutdown documentation, but if it is there (did not search exhaustively) it would be good to update as part of this PR.

dakrone commented on server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java:
```diff
@@ -366,8 +394,17 @@ public AllocationStatus getLastAllocationStatus() {
+            .filter(shutdownMetadata -> SingleNodeShutdownMetadata.Type.RESTART.equals(shutdownMetadata.getType()))
+            .map(SingleNodeShutdownMetadata::getShardReallocationDelay)
+            .map(TimeValue::nanos)
+            .orElse(INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.get(indexSettings).nanos());
```
> That seems wrong? I think we should take the greater of the shutdown delay and the delayed allocation delay.

I think it should be the explicitly set value in the node shutdown, regardless of the cluster-level setting for index delayed node left timeout. In that way we are doing exactly what the user asked (you asked me to wait 10 minutes, so I will wait 10 minutes regardless of the other setting) which is more expected. Why do you think it should be the max? That feels a little like action-at-a-distance when a user could not even be aware of the other cluster-wide setting.

henningandersen replied:
`index.unassigned.node_left.delayed_timeout` is an index level setting. The shutdown allocation delay is less likely to be controllable by a user (I believe cloud will simply use our default value) and I think of it as a "default value" for the index specific delay, allowing some extra (but never less) leeway to handle a planned restart.

If a user setup `index.unassigned.node_left.delayed_timeout=15m` for some indices, it seems odd that a planned restart would use a smaller value for the specific index/shards.

Also, if the shutdown API was called with `allocation_timeout=0`, it would be somewhat strange to disregard the index specific delay. That would effectively mean that we would be treating a planned restart as a worse event than a crash.

I think of restarts as something happening automatically, based on autoscaling, upgrades or other maintenance (though it could be an admin doing this, but without discussing with users). So if users (who should control the index level settings) set up their indices with higher delays than the shutdown uses, I think that higher delay should be used.

gwbrown replied:
I've changed this to use the max of index level delay and restart delay.

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/SingleNodeShutdownMetadata.java:
```diff
@@ -51,16 +56,24 @@
         return PARSER.apply(parser, null);
     }
 
+    public static final TimeValue DEFAULT_RESTART_SHARD_ALLOCATION_DELAY = TimeValue.timeValueMinutes(10);
```
We did actually discuss this in the meeting. Cloud uses a 5 minute timeout and we wanted something a bit higher than that and settled on 10 minutes. Is there a reason you think this should not be 10 minutes?

henningandersen replied:
Nothing very substantial, it just feels like a long time to expect a restart to take. Our documentation speaks of 5 minutes too. Our default delayed allocation is 1 minute. And if cloud uses 5 minute timeout, 10 minutes seems excessive.

We should remember that once the timeout is hit, nothing really bad happens, we "just" start doing more work than absolutely necessary in order to restore availability.

If cloud uses a 5 minute timeout, I would advocate using the same timeout here.

gwbrown replied:
I've changed the default value to `5m`.

AthenaEryma committed:
Use raw allocation delay value in equals/hashcode

AthenaEryma committed:
Adjust `delayed` calculation per review to...

1. Not ignore the default timeout for non-RESTART shutdowns
2. Ensure delayed=false if both timeouts are 0

gwbrown commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -469,10 +499,14 @@ private void disassociateDeadNodes(RoutingAllocation allocation) {
             for (ShardRouting shardRouting : node.copyShards()) {
                 final IndexMetadata indexMetadata = allocation.metadata().getIndexSafe(shardRouting.index());
-                boolean delayed = INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.get(indexMetadata.getSettings()).nanos() > 0;
+                boolean delayed = Optional.ofNullable(nodesShutdownMetadata.get(node.nodeId()))
+                    // If we know this node is restarting, then the allocation should be delayed
+                    .map(shutdownMetadata -> SingleNodeShutdownMetadata.Type.RESTART.equals(shutdownMetadata.getType()))
+                    // Otherwise, use the "normal" allocation delay logic
+                    .orElse(INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.get(indexMetadata.getSettings()).nanos() > 0);
```
> I think this means that if the shutdown type is REMOVE or REPLACE, we now ignore the delayed left timeout?

Good catch. I'll fix.

> Also, I wonder if we should only use the shutdown indication if it had completed, i.e., the node node was ready to restart when this happened?

We don't actually keep state about whether the shutdown preparation is complete or not, we compute the status when the Get Status API is called. So we'd have to move that logic to somewhere where it's accessible here, and ensure it always remains fast since this is called on the cluster state update thread.

It would likely be better if we checked the status, I agree, but RESTART-type shutdowns should be relatively quick to prepare. Do you think this is something we need for v1, or could we introduce it as the shutdown process becomes more involved?

> A flapping node during a restart could result in a long period of not assigning the shard, since we recalculate the delay every time the shard is recovered and the node subsequently dies.

Can't this happen today with the existing allocation delay? Admittedly, it's less likely with a shorter timeout.

gwbrown replied:
There's also the case of someone simply restarting the node before it's ready. Should we use the general delay in that case? It's not clear to me that one behavior is more intuitive than the other.

henningandersen replied:
> Do you think this is something we need for v1, or could we introduce it as the shutdown process becomes more involved?

I do not think it is mandatory for v1.

> Can't this happen today with the existing allocation delay? Admittedly, it's less likely with a shorter timeout.

This was specifically for the case where "they setup no delay for their indices". We would be allocating those immediately. Obviously during a cloud restart today, there would be no such allocation going on though.

Given the technical difficulty in doing this, we can defer to a follow-up.

henningandersen replied:
> There's also the case of someone intentionally restarting the node before it's ready. Should we use the general delay in that case? It's not clear to me that one behavior is more intuitive than the other.

Thanks, good point. I suppose that for a `RESTART` shutdown, the node is really "ready" for restart immediately (since we always anticipate a restart), however, optimally cloud would wait a little before forcing it through. So relying on the presence of the shutdown indication seems good here.

gwbrown replied:
I've changed this to track whether we knew the last allocated node was restarting at the time the shard became unassigned, and if not, to always use the index-level delay rather than the restart-level delay.

AthenaEryma committed:
Use randomFrom(Type.value()) to generate test shutdown metadata per review

AthenaEryma committed:
Use `restartNode()` instead of explicitly stopping/starting per review

gwbrown commented on x-pack/plugin/shutdown/src/internalClusterTest/java/org/elasticsearch/xpack/shutdown/NodeShutdownDelayedAllocationIT.java:
```diff
@@ -0,0 +1,230 @@
+        AcknowledgedResponse putShutdownResponse = client().execute(PutShutdownNodeAction.INSTANCE, putShutdownRequest).get();
+        assertTrue(putShutdownResponse.isAcknowledged());
+
+        // Actually stop the node
+        internalCluster().stopRandomNode(InternalTestCluster.nameFilter(nodeToRestartName));
+
+        // Verify that the shard's allocation is delayed
+        assertBusy(() -> { assertThat(client().admin().cluster().prepareHealth().get().getDelayedUnassignedShards(), equalTo(1)); });
+
+        // Bring the node back
+        internalCluster().startNode(nodeToRestartDataPathSettings); // this will use the same data location as the stopped node
```
TIL about `RestartCallback`, thanks.

AthenaEryma committed:
Clean up timeouts per review

gwbrown commented on x-pack/plugin/shutdown/src/internalClusterTest/java/org/elasticsearch/xpack/shutdown/NodeShutdownDelayedAllocationIT.java:
```diff
@@ -0,0 +1,230 @@
+        ensureGreen(TimeValue.timeValueSeconds(30), "test");
+    }
+
+    public void testShardAllocationStartsImmediatelyIfShutdownDeleted() throws Exception {
```
I'll update this to go with option 1, I don't like relying on randomization to check actually-different code paths like that.

AthenaEryma committed:
Factor out shared test logic per review

AthenaEryma committed:
Add type & allocation delay validation to PutShutdownNodeAction.Request per review

AthenaEryma committed:
Change delay field to `allocation_delay`

AthenaEryma committed:
spotless

gwbrown commented:
Hm, I appear to have oops'd the commit history. Please stand by.

gwbrown commented:
@henningandersen Docs for this feature are in a separate PR, https://github.com/elastic/elasticsearch/pull/74505. I'll be sure those are updated shortly.

AthenaEryma committed:
Merge branch 'master' into decom/delayed-shard-reassignment-on-restart

AthenaEryma committed:
Change default delay when restarting to 5m

AthenaEryma committed:
Use maximum of `index-level delay` and `restart delay`.

AthenaEryma committed:
Determine delay based on shutdown state at time shard became unassigned

AthenaEryma committed:
Merge branch 'master' into decom/delayed-shard-reassignment-on-restart

gwbrown commented:
@henningandersen I believe I've addressed all of your feedback - thanks again for the input!

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java:
```diff
@@ -242,8 +266,12 @@ public UnassignedInfo(Reason reason, @Nullable String message, @Nullable Excepti
         this.failedNodeIds = Collections.unmodifiableSet(failedNodeIds);
-        assert (failedAllocations > 0) == (reason == Reason.ALLOCATION_FAILED) :
-            "failedAllocations: " + failedAllocations + " for reason " + reason;
+        this.lastAllocatedNodeId = lastAllocatedNodeId;
```
can we assert that `lastAllocatedNodeIsRestarting == false || lastAllocatedNodeId != null`?

gwbrown replied:
I've added such a check.

henningandersen commented on server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java:
```diff
@@ -99,11 +101,14 @@ public void processExistingRecoveries(RoutingAllocation allocation) {
                                 nodeWithHighestMatch + "]",
                             null, 0, allocation.getCurrentNanoTime(), System.currentTimeMillis(), false,
-                            UnassignedInfo.AllocationStatus.NO_ATTEMPT, failedNodeIds);
+                            UnassignedInfo.AllocationStatus.NO_ATTEMPT, failedNodeIds, null, nodeIsRestarting);
```
I think we should just pass in false for restarting here?

Also, I wonder if we should for completeness pass in the `currentNodeId` for `lastAllocatedNodeId`? Not that it truly matters I think.

gwbrown replied:
To the first point, this boolean no longer exists.

To the second, we cannot due so due to the assertion recommended above that `lastAllocatedNodeId` only be populated when the node has left (and here it has a different reason).

henningandersen commented on x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/PutShutdownNodeAction.java:
```diff
@@ -101,6 +114,10 @@ public ActionRequestValidationException validate() {
             }
 
+            if (shardReallocationDelay != null && SingleNodeShutdownMetadata.Type.RESTART.equals(type) == false) {
+                arve.addValidationError(ALLOCATION_DELAY_FIELD + "is only allowed for RESTART-type shutdown requests");
```
```suggestion
                arve.addValidationError(ALLOCATION_DELAY_FIELD + " is only allowed for RESTART-type shutdown requests");
```

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java:
```diff
@@ -533,9 +538,21 @@ assert getByAllocationId(failedShard.shardId(), failedShard.allocationId().getId
+                            0,
+                            unassignedInfo.getUnassignedTimeInNanos(),
+                            unassignedInfo.getUnassignedTimeInMillis(),
+                            false,
```
It is a bit annoying that we have to calculate `nodeIsRestarting` when we pass in `false` to delayed anyway. I wonder if we were better off by making delayed an enum with values "NORMAL", "SHUTDOWN" (and possibly use "NO" instead of null)? That would avoid calculating `nodeIsRestarting` above for no good reason.

gwbrown replied:
This no longer needs to be calculated here, as the reason is `PRIMARY_FAILED`.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java:
```diff
@@ -366,8 +420,20 @@ public AllocationStatus getLastAllocationStatus() {
+            .map(nodesShutdownMap::get)
+            .filter(shutdownMetadata -> SingleNodeShutdownMetadata.Type.RESTART.equals(shutdownMetadata.getType()))
+            .map(SingleNodeShutdownMetadata::getShardReallocationDelay)
+            .map(TimeValue::nanos)
```
nit: The auto-boxing here annoys me slightly, I would prefer to terminate the optional map'ing here and work on the longs. We need Optional.mapToLong :slightly_smiling_face:. I can live with it as is though, escape analysis might catch it and otherwise it is unlikely to be a big deal (I think).

gwbrown replied:
Given that it 1) gets immediately auto-unboxed back to a primitive `long` and 2) is not possible to become null before it gets unboxed, I'm not sure what we lose here other than an unnecessary object allocation or two? (given that it'll be unboxed for the `Math.max` call and reboxed into the optional afterwards)

And it seems to me the increase in readability outweighs the slight inefficiency.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -207,9 +210,12 @@ public ClusterState applyFailedShards(final ClusterState clusterState, final Lis
+                    .map(shutdownInfo -> shutdownInfo.getType().equals(Type.RESTART))
+                    .orElse(false);
                 UnassignedInfo unassignedInfo = new UnassignedInfo(UnassignedInfo.Reason.ALLOCATION_FAILED, message,
                     failedShardEntry.getFailure(), failedAllocations + 1, currentNanoTime, System.currentTimeMillis(), false,
```
This is another case where we calculate `nodeIsRestarting` without really needing it.

gwbrown replied:
Not anymore!

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -300,13 +306,28 @@ private void removeDelayMarkers(RoutingAllocation allocation) {
+                            unassignedInfo.getLastAllocationStatus(),
+                            unassignedInfo.getFailedNodeIds(),
+                            unassignedInfo.getLastAllocatedNodeId(),
+                            unassignedInfo.isLastAllocatedNodeIsRestarting()),
```
Also related to introducing a delayed enum, I would find it slightly nicer also for this case, since the boolean is unused when `delayed=false`.

gwbrown replied:
And this is taken care of by using `reason` instead.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -469,10 +503,17 @@ private void disassociateDeadNodes(RoutingAllocation allocation) {
+                    .orElse(false);
+                boolean delayed = delayedDueToKnownRestart
+                    || INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.get(indexMetadata.getSettings()).nanos() > 0;
+                boolean nodeIsRestarting = Optional.ofNullable(nodesShutdownMetadata.get(shardRouting.currentNodeId()))
```
I think we should always have `delayedDueToKnownRestart == nodeIsRestarting` so we can hopefully do just one of these calculations?

gwbrown replied:
This isn't 100% resolved - we still have to decide if the `reason` is `NODE_LEFT` or `NODE_RESTARTING`, which is separate from the `delayedDueToKnownRestart` because `delayed` should be `false` even if the node is restarting if the delay would be zero (as you pointed out in your first review).

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -469,10 +503,17 @@ private void disassociateDeadNodes(RoutingAllocation allocation) {
+                boolean nodeIsRestarting = Optional.ofNullable(nodesShutdownMetadata.get(shardRouting.currentNodeId()))
+                    .map(shutdownInfo -> shutdownInfo.getType().equals(Type.RESTART))
+                    .orElse(false);
                 UnassignedInfo unassignedInfo = new UnassignedInfo(UnassignedInfo.Reason.NODE_LEFT, "node_left [" + node.nodeId() + "]",
```
Thinking more about the delayed enum, I think we could consider adding a new `UnassignedInfo.Reason` signalling that this was a restart (to the best of our knowledge).

gwbrown replied:
This is what I went with, rather than changing `delayed` to an enum.

henningandersen reviewed (commented):
Thanks for working on this Gordon. I left a number of comments, but the bigger one is around the restarting flag. I think moving it to be a `Reason` would be ideal, since this could even be helpful when diagnosing clusters.

Seeing an unassigned shard with reason `NODE_LEFT` could means some system issue whereas if we can determine it is a `RESTART` (or very likely is one), it is sort of a non-issue.

I am OK with that determination being done like you do here, i.e., a node crashing during a restart will also look like a reason restart. It should be rare and we can likely iterate on it too.

AthenaEryma committed:
Switch to NODE_RESTARTING enum value instead of separate boolean

AthenaEryma committed:
Merge branch 'master' into decom/delayed-shard-reassignment-on-restart

AthenaEryma committed:
Rename to `allocationDelay`

AthenaEryma committed:
Add assert per review

AthenaEryma committed:
Add space to error message

gwbrown commented on server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java:
```diff
@@ -339,6 +378,21 @@ public String getDetails() {
+    /**
+     * Whether the node this shard was last assigned to was restarting when it became unassigned.
+     */
+    public boolean isLastAllocatedNodeIsRestarting() {
```
This method no longer exists as we capture this information the `reason` field.

gwbrown commented on server/src/main/java/org/elasticsearch/cluster/metadata/SingleNodeShutdownMetadata.java:
```diff
@@ -115,12 +134,27 @@ public long getStartedAtMillis() {
+     * reassigned while the node is restarting. Will be {@code null} for non-restart shutdowns.
+     */
+    @Nullable
+    public TimeValue getShardReallocationDelay() {
```
Done, I meant to do that everywhere when I changed some of them, but must have dropped it somewhere along the line.

AthenaEryma committed:
Merge branch 'master' into decom/delayed-shard-reassignment-on-restart

AthenaEryma committed:
Ensure we don't try to set lastNodeId if the reason isn't NODE_LEFT or NODE_RESTARTING

gwbrown commented:
@henningandersen Now that the tests are passing again, what do you think of the new approach using the reason enum?

henningandersen commented on server/src/test/java/org/elasticsearch/cluster/routing/UnassignedInfoTests.java:
```diff
@@ -290,23 +307,168 @@ public void testFailedShard() {
+
+        // Generate a random time value - but don't use nanos as extremely small values of nanos can break assertion calculations
+        final TimeValue shutdownDelay = TimeValue.parseTimeValue(
+            randomTimeValue(2, 1000, "d", "h", "ms", "s", "m", "micros"),
```
To avoid risking too many iterations on the `randomValueOtherThanMany` I propose:

```suggestion
            randomTimeValue(100, 1000, "d", "h", "ms", "s", "m", "micros"),
```

henningandersen commented on x-pack/plugin/shutdown/src/internalClusterTest/java/org/elasticsearch/xpack/shutdown/NodeShutdownDelayedAllocationIT.java:
```diff
@@ -0,0 +1,235 @@
+            nodeToRestartId,
+            SingleNodeShutdownMetadata.Type.RESTART,
+            this.getTestName(),
+            TimeValue.timeValueSeconds(1)
```
Why not just 1 millisecond?

henningandersen commented on server/src/test/java/org/elasticsearch/cluster/routing/UnassignedInfoTests.java:
```diff
@@ -76,10 +82,20 @@ public void testSerialization() throws Exception {
         int failedAllocations = randomIntBetween(1, 100);
```
I wonder if we should add a test of bwc serialization? Since the functionality will only be enabled in 7.15 and I am unaware of any shutdown rolling upgrade tests, we could in theory be exposed to a bwc bug in serialization (not that I saw any, looks all good to me).

gwbrown replied:
I'm not sure it's worth adding a unit test for this, but I do intend to add some rolling and full-restart upgrade tests for this soon.

henningandersen reviewed (approved):
LGTM. Thanks for the extra iterations. I wonder if @dakrone should have a quick re-review given that it changed substantially.

AthenaEryma committed:
Expand comment per review

AthenaEryma committed:
Modify assert to require lastAllocatedNodeId for NODE_RESTARTING

Rather than only allowing NODE_LEFT/NODE_RESTARTING to have
a lastAllocatedNodeId

AthenaEryma committed:
Revert "Ensure we don't try to set lastNodeId if the reason isn't NODE_LEFT or NODE_RESTARTING"

This reverts commit 8bc7442b96dc909532d79ae81e5d268a41aa1776.

gwbrown commented on server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:
```diff
@@ -719,6 +719,12 @@ public IndexMetadata getIndexSafe(Index index) {
             .orElse(Collections.emptyMap());
     }
 
+    public Map<String, SingleNodeShutdownMetadata> nodeShutdowns() {
```
No, there's a couple things that prevent this:
1) In `TransportPutShutdownNodeAction` when we get a `PUT` for a node that already has a record, it's updated rather than added to, and
2) The data structure used to store the `SingleNodeShutdownMetadata` (the `Map` in the line you're commenting on) is keyed by node UUID, so it should be impossible to have multiple records for the same key.

Since the node id is duplicated in the `SingleNodeShutdownMetadata` as well, it's conceivable that in the case of a bug we could end up with a mismatch between the id used for keying and the id used in the object, but I don't think that's likely.

henningandersen replied:
Yeah, sorry if this was unclear, what I meant was whether it could be a reasonable use case to have both a RESTART and one of the two others at the same time. Not that there is anything wrong in this PR or anything, more wanted to bring this to your attention for possible discussion (and maybe you discussed it already and discarded the use case?).

gwbrown commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -469,10 +506,25 @@ private void disassociateDeadNodes(RoutingAllocation allocation) {
-                UnassignedInfo unassignedInfo = new UnassignedInfo(UnassignedInfo.Reason.NODE_LEFT, "node_left [" + node.nodeId() + "]",
-                    null, 0, allocation.getCurrentNanoTime(), System.currentTimeMillis(), delayed, AllocationStatus.NO_ATTEMPT,
-                    Collections.emptySet());
+                boolean delayedDueToKnownRestart = Optional.ofNullable(nodesShutdownMetadata.get(node.nodeId()))
```
Good catch, thanks!

AthenaEryma committed:
Compute reason once per node when disassociating dead nodes, instead of per shard

AthenaEryma committed:
Remove unused code block per review

AthenaEryma committed:
Adjust UnassignedInfo test instance creation per review

AthenaEryma committed:
Adjust test delays per review

AthenaEryma committed:
imports

AthenaEryma committed:
Assert that NODE_RESTARTING requires lastAllocatedNodeId + fix tests

A number of tests created `UnassignedInfo`s without regard for the `reason`,
which started failing after adjusting the assertion. These tests now generate
their `UnassignedInfo`s via methods which handle the requirements.

AthenaEryma committed:
Merge branch 'master' into decom/delayed-shard-reassignment-on-restart

AthenaEryma committed:
Alter test to generated `lastAllocatedNodeId` if necessary

gwbrown commented:
Per brief discussion elsewhere, I'm going to go ahead and merge this - @dakrone can review the recent changes when he's back online, and any changes can be made as a follow-up.
